nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/CoreSession.java

```diff
@@ -1150,13 +1164,109 @@ DocumentModel restoreToVersion(DocumentRef docRef, DocumentRef versionRef, boole
     // LDAP/database.)
     List<String> getAvailableSecurityPermissions();
 
+    /**
+     * Turns the document into a record.
+     * <p>
+     * A record is a document with specific capabilities related to mandatory retention until a given date, and legal
+     * holds. In addition, its main blob receives special treatment from the document blob manager to make sure it's
+     * never shared with another blob at the storage level, and is deleted as soon as the record is deleted.
+     * <p>
+     * If the document is already a record, this method has no effect.
+     * <p>
+     * The permission {@value org.nuxeo.ecm.core.api.security.SecurityConstants#CAN_MAKE_RECORD} is required.
+     *
+     * @param docRef the document
+     * @see #isRecord
+     * @since 11.1
+     */
+    void makeRecord(DocumentRef docRef);
+
+    /**
+     * Checks if the document is a record.
+     *
+     * @param docRef the document
+     * @return {@code true} if the document is a record, {@code false} otherwise
+     * @see #makeRecord
+     * @since 11.1
+     */
+    boolean isRecord(DocumentRef docRef);
+
+    /**
+     * The special date that corresponds to a retention date in the indeterminate future.
+     *
+     * @see #setRetentionUntil
+     * @since 11.1
+     */
+    Calendar RETENTION_UNTIL_INDETERMINATE = GregorianCalendar.from(ZonedDateTime.parse("9999-01-01T00:00Z"));
+
+    /**
+     * Sets a retention date to the given value.
+     * <p>
+     * If no previous retention date was set, or if the previous retention date was
+     * {@linkplain #RETENTION_UNTIL_INDETERMINATE indeterminate}, or if the previous retention date was <em>before</em>
+     * the given value, then the retention date is set to the given value.
+     * <p>
+     * If the previous retention date was <em>after</em> the given value (that is, if trying to reduce the retention
+     * time), an exception is thrown.
+     * <p>
+     * If the given value is {@code null} and the previous retention date is in the past (it has already expired), then
```

Comment:
Special treatment about `null` shouldn't be a low level action (low level `Session`)?
Because retention date will be removed by an async processing and probably never be done through a user action (or I'm missing something).